src/arcsf/eval/evaluate_model.py

```diff
@@ -0,0 +1,126 @@
+import argparse
+import json
+import os
+
+import numpy as np
+import torch
+import transformers
+import yaml
+from transformers import AutoModelForCausalLM, AutoTokenizer
+
+from arcsf.data.data_module import BlankQAFormatter, EvalQADataset, get_data
+from arcsf.eval.utils import all_eval, combine_dicts, get_metrics
+from arcsf.utils import get_device
+
+
+def evaluate_model(
+    model: torch.nn.Module,
+    base_truth_ratios_path: str,
+    tokenizer: transformers.AutoTokenizer,
+    experiment_config: dict,
+    **generate_kwargs: dict,
+) -> dict[str, float]:
+    """
+    Evaluates the model against a baseline model, given the model, tokenizer, experiment
+    config and the path to the baseline model truth ratios.
+
+    Args:
+        model: Model to perform the evaluation on
+        base_truth_ratios_path: path for the baseline mode truth ratios
+        tokenizer: tokenizer being used for the model
+        experiment_config: experiment config (currently a dictionary)
+
+    Returns:
+        dictionary of aggregated metrics for evaluation (generated using `get_metrics`)
+    """
+    random_seed = experiment_config["seed"]
+
+    # get datasets
+    forget_data, retain_data = get_data(
+        "tofu", random_seed=random_seed, **experiment_config["data_config"]
+    )
+    # get available device
+    device = get_device()
+
+    # create datasets
+    retain_dataset = EvalQADataset(
+        retain_data,
+        tokenizer,
+        BlankQAFormatter(),
+        "standard",
+        quantitative_eval=True,
+        qualitative_eval=True,
+        device=device,
+        random_seed=random_seed,
+    )
+    forget_dataset = EvalQADataset(
+        forget_data,
+        tokenizer,
+        BlankQAFormatter(),
+        "standard",
+        quantitative_eval=True,
+        qualitative_eval=True,
+        device=device,
+        random_seed=random_seed,
+    )
+
+    # perform evaluation on models to get the raw values
+    retain_values = all_eval(
+        model,
+        retain_dataset,
+        1,
+        device,
+        tokenizer,
+        **generate_kwargs,
+    )
+    forget_values = all_eval(
+        model,
+        forget_dataset,
+        1,
+        device,
+        tokenizer,
+        **generate_kwargs,
+    )
+
+    # combine dictionaries and load the baseline model truth ratios
+    analysis_vals = combine_dicts(forget_values, retain_values)
+    base_truth_ratios = torch.tensor(np.loadtxt(base_truth_ratios_path))
```

Comment:
I've not checked the paper as of writing this, but shouldn't this be using base truth ratios from the retain data and the forget data here (rather than just one set of truth ratios)?

Reply:
The base model truth ratios are only used in the forget quality, so we only need them from the forget set. We could load from the retain set as well for completeness, but I don't think they'll be used our the current implementation.